orttraining/orttraining/training_ops/cuda/optimizer/adam/adam_impl.cu

```diff
@@ -0,0 +1,221 @@
+#include "core/providers/cuda/cuda_common.h"
+#include "core/providers/cuda/cu_inc/common.cuh"
+#include "orttraining/training_ops/cuda/optimizer/common.cuh"
+#include "orttraining/training_ops/cuda/optimizer/common.h"
+#include "core/providers/cuda/cu_inc/common.cuh"
+namespace onnxruntime {
+namespace cuda {
+
+template <typename T_WEIGHT, typename T_GRAD, typename T_MOMENTUM>
+__device__ void PrepareMTAData(
+    const ChunkGroup<MTA_ADAM_GROUP_SIZE>& chunks,
+    const int& block_idx,
+    T_WEIGHT*& weight_chunk_ptr,
+    T_GRAD*& grad_chunk_ptr,
+    T_MOMENTUM*& momentum_1_chunk_ptr,
+    T_MOMENTUM*& momentum_2_chunk_ptr,
+    int& chunk_size) {
+  const int tensor_idx = chunks.block_index_to_tensor_group_index[block_idx];
+  const int tensor_size = chunks.tensor_sizes[tensor_idx];
+  T_WEIGHT* weight_tensor_ptr = static_cast<T_WEIGHT*>(chunks.tensor_ptrs[0][tensor_idx]);
+  T_GRAD* grad_tensor_ptr = static_cast<T_GRAD*>(chunks.tensor_ptrs[1][tensor_idx]);
+  T_MOMENTUM* momentum_1_tensor_ptr = static_cast<T_MOMENTUM*>(chunks.tensor_ptrs[2][tensor_idx]);
+  T_MOMENTUM* momentum_2_tensor_ptr = static_cast<T_MOMENTUM*>(chunks.tensor_ptrs[3][tensor_idx]);
+  const int chunk_start_idx = chunks.block_index_to_chunk_start_index[block_idx];
+  // chunk_size is chunks.chunk_size if the loaded chunk is full. Otherwise (this
+  // chunk is the last one in the source tensor), the actual size is determined
+  // by the bound of the source tensor.
+  chunk_size = min(tensor_size, chunk_start_idx + chunks.chunk_size) - chunk_start_idx;
+
+  weight_chunk_ptr = weight_tensor_ptr + chunk_start_idx;
+  grad_chunk_ptr = grad_tensor_ptr + chunk_start_idx;
+  momentum_1_chunk_ptr = momentum_1_tensor_ptr + chunk_start_idx;
+  momentum_2_chunk_ptr = momentum_2_tensor_ptr + chunk_start_idx;
+}
+
+// Torch Adam equivalence.
+template <typename T_WEIGHT, typename T_GRAD, typename T_MOMENTUM>
+__global__ void AdamComputeMode0(
+    ChunkGroup<MTA_ADAM_GROUP_SIZE> chunks,
+    const float alpha,
+    const float beta,
+    const float epsilon,
+    const float lr,
+    const float alpha_correction,
+    const float beta_correction,
+    const float decay) {
+  const int block_idx = blockIdx.x;
+
+  T_WEIGHT* weight_chunk_ptr;
+  T_GRAD* grad_chunk_ptr;
+  T_MOMENTUM* momentum_1_chunk_ptr;
+  T_MOMENTUM* momentum_2_chunk_ptr;
+
+  // TODO: unroll this one for better perf.
+  int chunk_size;
+
+  PrepareMTAData(chunks, block_idx, weight_chunk_ptr, grad_chunk_ptr,
+                 momentum_1_chunk_ptr, momentum_2_chunk_ptr, chunk_size);
+
+  // A shared constant.
+  const float one = 1.0f;
+#pragma unroll(4)
+  for (int i = threadIdx.x; i < chunk_size; i += blockDim.x) {
+    float w = static_cast<float>(weight_chunk_ptr[i]);
+    float g = static_cast<float>(grad_chunk_ptr[i]);
+    float m1 = static_cast<float>(momentum_1_chunk_ptr[i]);
+    float m2 = static_cast<float>(momentum_2_chunk_ptr[i]);
+
+    // Perform weight decay.
+    w = w - (w * lr * decay);
```

Comment:
It seems like this implementation is the `AdamW` implementation (as referred to in other libraries) due to the weight decay being applied on the weight instead of the gradient.
Shall we make it explicit by calling the op (and the kernel) `AdamW`?

Reply:
Good suggestion.

Reply:
I will update separately. Thanks.